packages/waku/src/router/client.ts

```diff
@@ -31,13 +31,37 @@ import {
   LOCATION_ID,
 } from './common.js';
 import type { RouteProps, ShouldSkip } from './common.js';
+import type { createPages } from './create-pages.js';
+import type { CollectPaths } from './inferred-router-types.js';
 
 declare global {
   interface ImportMeta {
     readonly env: Record<string, string>;
   }
 }
 
+type AnyRouter = ReturnType<typeof createPages>;
+
+type RouterSpec<Router extends AnyRouter> = Awaited<
+  NonNullable<Router['DO_NOT_USE_ready']>
+>[number];
+
+// eslint-disable-next-line @typescript-eslint/no-empty-object-type
+export interface Register {
+  // router to be overridden by users
+}
+
+type RegisteredRouter = Register extends {
+  router: infer UserRouter extends AnyRouter;
+}
+  ? UserRouter
+  : AnyRouter;
+
+export type InferredPaths =
+  CollectPaths<RouterSpec<RegisteredRouter>> extends never
+    ? string
+    : CollectPaths<RouterSpec<RegisteredRouter>>;
```

Comment:
Can we just allow overriding the final result?

Reply:
I'd imagine that the most common use case would be to use the type inferred from the router.

That being said, I can add support to loosen the final result. I can think of two ways a user might want to do that:

1. Use inferred router page paths for autocomplete, but allow any string `type Path = '/' | (string & {})`
2. Full override the final result entirely to define whatever the user wants as their paths distinct from what goes through the Router `type Path = '/' | '/about'`

Reply:
I'm not trying to say to loosen the result.

Reply:
What are you hoping to do with overriding the final result then?

Reply:
Resolving this... going to change to: https://github.com/dai-shi/waku/pull/854#discussion_r1749041657